verification.py

```diff
@@ -4399,137 +4569,101 @@ def _run_pyvolve_comparison(self, model, length=20, num_samples=10):
         df_py, df_ts, alleles = self._run_pyvolve_stats(model, length, num_samples)
         self.plot_stats(df_py, df_ts, alleles, "pyvolve", model)
 
-    def add_pyvolve(self, model):
-        key = f"pyvolve-{model}"
 
-        def f():
-            self._run_pyvolve_comparison(model)
+@attr.s
+class PyvolveTest1(PyvolveTest):
+    model = attr.ib(type=str)
+
+    def run(self, output_dir):
+        self._output_dir = output_dir
+        logging.info("running", self.group, self.name)
+        self._run_pyvolve_comparison(self.model)
 
-        self._instances[key] = f
 
-    def add_instances(self):
-        self.add_pyvolve("JC69")
-        self.add_pyvolve("HKY")
+def register_pyvolve_tests(runner):
+    def register(model):
+        group = "pyvolve"
+        name = f"{group}_{model}"
+        runner.register(PyvolveTest1(name=name, group=group, model=model))
 
+    register("JC69")
+    register("HKY")
 
-def get_test_names():
-    tests = []
 
-    def get_class_name(c):
-        return f"{c}"[17:-6].lower()
+@attr.s
+class TestRunner:
+    """
+    Class responsible for registering all known tests and running
+    them.
+    """
+
+    tests = attr.ib(init=False, default=[])
+    groups = attr.ib(init=False, default=set())
+
+    def register(self, test):
+        self.tests.append(test)
+        self.groups.add(test.group)
 
-    # use a simple DFS to get the names of classes and
-    classes = SimulationVerifier.__subclasses__()
-    while classes:
-        top = classes.pop()
-        tests.append(get_class_name(top))
-        for c in top.__subclasses__():
-            classes.append(c)
-    return tests
+    def run(self, output_dir, names=None, group=None):
+        if names != []:
```

Comment:
It would probably be simplest to say that ``names`` and ``group`` are mutually exclusive here, so we have something like
```python
if names is not None and groupd is not None:
     raise ValueError("Cannot specify test names and group at the same time")
if names is not None:
     tests = [test for test in self.tests if test.name in names]
elif group is not None:
    tests = [test for test in self.tests if test.group == group]
else:
    tests = self.tests

￼```

This way we're just reasoning about the default values.